crates/rustc_codegen_spirv/src/builder/intrinsics.rs

```diff
@@ -418,6 +414,78 @@ impl Builder<'_, '_> {
             _ => self.fatal("count_ones on a non-integer type"),
         }
     }
+    pub fn bit_reverse(&self, arg: SpirvValue) -> SpirvValue {
+        let ty = arg.ty;
+        match self.cx.lookup_type(ty) {
+            SpirvType::Integer(bits, signed) => {
+                let u32 = SpirvType::Integer(32, false).def(self.span(), self);
+                let uint = SpirvType::Integer(bits, false).def(self.span(), self);
+
+                match (bits, signed) {
+                    (8 | 16, signed) => {
+                        let arg = arg.def(self);
+                        let arg = if signed {
+                            self.emit().bitcast(uint, None, arg).unwrap()
+                        } else {
+                            arg
+                        };
+                        let arg = self.emit().u_convert(u32, None, arg).unwrap();
+
+                        let reverse = self.emit().bit_reverse(u32, None, arg).unwrap();
+                        let shift = self.constant_u32(self.span(), 32 - bits).def(self);
+                        let reverse = self.emit().shift_right_logical(u32, None, reverse, shift).unwrap();
+                        let reverse = self.emit().u_convert(uint, None, reverse).unwrap();
+                        if signed {
+                            self.emit().bitcast(ty, None, reverse).unwrap()
+                        } else {
+                            reverse
+                        }
+                    }
+                    (32, false) => self.emit().bit_reverse(u32, None, arg.def(self)).unwrap(),
+                    (32, true) => {
+                        let arg = self.emit().bitcast(u32, None, arg.def(self)).unwrap();
+                        let reverse = self.emit().bit_reverse(u32, None, arg).unwrap();
+                        self.emit().bitcast(ty, None, reverse).unwrap()
+                    },
+                    (64, signed) => {
+                        let u32_32 = self.constant_u32(self.span(), 32).def(self);
+                        let arg = arg.def(self);
+                        let lower = self.emit().s_convert(u32, None, arg).unwrap();
+                        let higher = self
+                            .emit()
+                            .shift_left_logical(ty, None, arg, u32_32)
```

Comment:
I think this should be `shift_right_logical` ?

Reply:
all the 64bit bitshifts were wrong :D